src/backend/metal/src/device.rs

```diff
@@ -1046,24 +1046,35 @@ impl hal::Device<Backend> for Device {
                         }
                     }
                 }
-                n::DescriptorSetLayout::ArgumentBuffer { ref bindings, .. } => {
-                    //Note: we specifically ignore the stage flags and unconditionally
-                    // add resource overrides and bump the counter, because there is no
-                    // way to override the argument buffer binding itself in SPIRV-Cross.
-                    //TODO: ^^
-                    for &mut (_stage_bit, stage, ref mut counters) in stage_infos.iter_mut() {
-                        res_overrides.extend(bindings
-                            .iter()
-                            .map(|(&binding, arg)| {
-                                let key = msl::ResourceBindingLocation {
-                                    stage,
-                                    desc_set: set_index as _,
-                                    binding,
-                                };
-                                (key, arg.res.clone())
-                            })
+                n::DescriptorSetLayout::ArgumentBuffer { ref bindings, stage_flags, .. } => {
+                    for &mut (stage_bit, stage, ref mut counters) in stage_infos.iter_mut() {
+                        let has_stage = stage_flags.contains(stage_bit);
+                        res_overrides.insert(
+                            msl::ResourceBindingLocation {
+                                stage,
+                                desc_set: set_index as _,
+                                binding: msl::ARGUMENT_BUFFER_BINDING,
+                            },
+                            msl::ResourceBinding {
+                                buffer_id: if has_stage { counters.buffers } else { !0 },
```

Comment:
Why is `counters.buffers` guaranteed to be correct here?

Reply:
We are at the stage of going through resources and bumping the counters. Before, we were first going though resources and then through the push constants. With #2840, we were basically bumping the buffer counters 1:1 with enumerating descriptor sets, which matched spirv-cross default mapping of argument buffers. As of #2857, we are scanning the push constants first, which broke the matching.

This PR uses the new overrides, so that we can force the mapping again.